src/deadline/job_attachments/upload.py

```diff
@@ -82,6 +83,50 @@
 
 logger = logging.getLogger("deadline.job_attachments.upload")
 
+
+class FileStatCache:
+    """Cache file stat results to avoid redundant filesystem calls"""
+
+    def __init__(self):
+        self._cache = {}
+
+    def get_stat(self, path: Path) -> Optional[os.stat_result]:
+        """Get cached stat result or perform stat and cache it"""
+        path_str = str(path)
+        if path_str not in self._cache:
+            try:
+                self._cache[path_str] = path.stat()
+            except (FileNotFoundError, PermissionError, OSError):
+                self._cache[path_str] = None
+        return self._cache[path_str]
```

Comment:
You could optimize this further using [`functools.lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache).

Regardless of that, we should consider is having an unbounded cache - since this should be a best-effort optimization and not cause out of memory crashes.

There was a recent issue (#849) related to this same concern.